tests/test_pipeline/test_schedule/test_zerobubble_pp.py

```diff
@@ -0,0 +1,734 @@
+            # microbatch 2
+            # chunk 0 fwd
+            ScheduledNode(type="RECV_FORWARD", chunk=0, stage=3, minibatch=2),
+            ScheduledNode(type="F", chunk=0, stage=3, minibatch=2),
+            ScheduledNode(type="SEND_FORWARD", chunk=0, stage=3, minibatch=2),
+            # chunk 1 fwd
+            ScheduledNode(type="RECV_FORWARD", chunk=1, stage=3, minibatch=2),
+            ScheduledNode(type="F", chunk=1, stage=3, minibatch=2),
+            ScheduledNode(type="SEND_FORWARD", chunk=1, stage=3, minibatch=2),
+            # chunk 1 bwd
+            ScheduledNode(type="RECV_BACKWARD", chunk=1, stage=3, minibatch=2),
+            ScheduledNode(type="B", chunk=1, stage=3, minibatch=2),
+            ScheduledNode(type="W", chunk=1, stage=3, minibatch=2),
+            ScheduledNode(type="SEND_BACKWARD", chunk=1, stage=3, minibatch=2),
+            # chunk 0 bwd
+            ScheduledNode(type="RECV_BACKWARD", chunk=0, stage=3, minibatch=2),
+            ScheduledNode(type="B", chunk=0, stage=3, minibatch=2),
+            ScheduledNode(type="W", chunk=0, stage=3, minibatch=2),
+            ScheduledNode(type="SEND_BACKWARD", chunk=0, stage=3, minibatch=2),
+            # microbatch 3
+            # chunk 0 fwd
+            ScheduledNode(type="RECV_FORWARD", chunk=0, stage=3, minibatch=3),
+            ScheduledNode(type="F", chunk=0, stage=3, minibatch=3),
+            ScheduledNode(type="SEND_FORWARD", chunk=0, stage=3, minibatch=3),
+            # chunk 1 fwd
+            ScheduledNode(type="RECV_FORWARD", chunk=1, stage=3, minibatch=3),
+            ScheduledNode(type="F", chunk=1, stage=3, minibatch=3),
+            ScheduledNode(type="SEND_FORWARD", chunk=1, stage=3, minibatch=3),
+            # chunk 1 bwd
+            ScheduledNode(type="RECV_BACKWARD", chunk=1, stage=3, minibatch=3),
+            ScheduledNode(type="B", chunk=1, stage=3, minibatch=3),
+            ScheduledNode(type="W", chunk=1, stage=3, minibatch=3),
+            ScheduledNode(type="SEND_BACKWARD", chunk=1, stage=3, minibatch=3),
+            # chunk 0 bwd
+            ScheduledNode(type="RECV_BACKWARD", chunk=0, stage=3, minibatch=3),
+            ScheduledNode(type="B", chunk=0, stage=3, minibatch=3),
+            ScheduledNode(type="W", chunk=0, stage=3, minibatch=3),
+            ScheduledNode(type="SEND_BACKWARD", chunk=0, stage=3, minibatch=3),
+        ],
+    ]
+
+    scheduler = ZeroBubbleVPipeScheduler(
+        schedule=zbv_schedule,  # hint: send whole schedule or local schedule only ?
+        stage_manager=stage_manager,
+        num_model_chunks=pp_size,
+        num_microbatch=num_microbatch,
+        overlap_p2p=False,
+    )
+
+    # loss func
+    def criterion(x, *args, **kwargs):
+        return (x * x).mean()
+
+    # init model and input
+    batch_size = 4
+    num_layers = 8
+    in_dim = out_dim = 8
+    print(f"Before init Model: {torch.cuda.memory_allocated()/1024**3 :.3f} GB on device {stage_manager.get_rank()};")
```

Comment:
replace it with assertation?

Reply:
pytest won't check stdout

Reply:
> replace it with assertation?

Add memory assertation in https://github.com/hpcaitech/ColossalAI/pull/6034/commits/2f09c374f3dda68fe3b5253ca7ba5df25323dd30;
As shown below, there will be minor differences between them. So I round(, 5) to them;
<img width="770" alt="32b7af1ae3f7962061c520b860c972c" src="https://github.com/user-attachments/assets/561cf13b-48b0-4a57-91b9-2a0c738d0529">